src/librustc_middle/mir/interpret/int.rs

```diff
@@ -0,0 +1,36 @@
+use rustc_target::abi::Size;
+
+#[derive(Copy, Clone)]
+/// A type for representing any integer. Only used for printing.
+// FIXME: Use this for the integer-tree representation needed for type level ints and
+// const generics?
+pub struct ConstInt {
+    size: u8,
+    signed: bool,
+    bits: u128,
+}
+
+impl ConstInt {
+    pub fn new(bits: u128, size: Size, signed: bool) -> Self {
+        Self { bits, size: size.bytes() as u8, signed }
+    }
+}
+
+impl std::fmt::Debug for ConstInt {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let Self { size, signed, bits } = *self;
+        match (size, signed) {
+            (1, false) => write!(fmt, "{}", bits as u8),
+            (2, false) => write!(fmt, "{}", bits as u16),
+            (4, false) => write!(fmt, "{}", bits as u32),
+            (8, false) => write!(fmt, "{}", bits as u64),
+            (16, false) => write!(fmt, "{}", bits as u128),
+            (1, true) => write!(fmt, "{}", bits as i8),
+            (2, true) => write!(fmt, "{}", bits as i16),
+            (4, true) => write!(fmt, "{}", bits as i32),
+            (8, true) => write!(fmt, "{}", bits as i64),
+            (16, true) => write!(fmt, "{}", bits as i128),
+            _ => bug!("ConstInt {} with size = {} and signed = {}", bits, size, signed),
+        }
```

Comment:
I seem to recall we already have some sign-aware pretty-printing of constants somewhere. Can we deduplicate that code (or open a cleanup issue to do so in the future)?

Reply:
done. Fun fallout :laughing: